### ISSUE TITLE
Request for 3.31 Patch Release (Code Insights Backend Worker)

### ISSUE THREAD

**Issue by Joelkw** [sourcegraph/sourcegraph#24489]
@sourcegraph/distribution I am requesting the following commits be included in a patch release. They are already merged into `main`:

* 7fcf7c4e5c176a04e4ebc7884cdac6abc31b4fd2 

> Are users/customers actively asking us for these changes and cannot wait until the next full release?

Yes – this issue effectively blocks the backend worker for code insights (in other words, it blocks insights over all repos as well as the ability to filter insights, a feature only available on insights over all repos). This particularly impacts the following customers and prospects, any one (or two) of which would be enough (IMO, to be fair) to justify the patch issue: 

* [$Customer1](https://docs.google.com/spreadsheets/d/1OEhzdMSlkGOaWyGKwdiAGlirsKKj9EN45Izn7kdKNTg/edit#gid=0&range=A4615) had a QBR **today** (8/31) asking about insights and hoping they could get their postgres instance unblocked. This customer is both actively trying to explore insights ([slack post from yesterday](https://sourcegraph.slack.com/archives/CTN0ZBETE/p1630361081004700)) and upgrading to 3.31 already ([slack post from today](https://sourcegraph.slack.com/archives/CTN0ZBETE/p1630426881005700)). We have been trying to build momentum at $Customer1 – who is one of our largest 10 customers – for 5+ months, blocked by their lack of upgrades – we do not want to lose the momentum we have now and the last they heard, code insights works over all their repos. 

* [$Customer2](https://docs.google.com/spreadsheets/d/1OEhzdMSlkGOaWyGKwdiAGlirsKKj9EN45Izn7kdKNTg/edit#gid=0&range=A1902) is planning to upgrade to 3.31 sometime this week, specifically so they can finally get insights over all their repos. We have previously worked with this customer but ran into issues (in various June/July posts [in slack](https://sourcegraph.slack.com/archives/CTGBWCKM0)) that 3.31 unblocks. The AE (Surdi) and CE (Christine) also have context here – their plans to update to 3.31 are in an email thread. This is one of our three most promising existing prospects for the insights feature based on their early adoption. 

* [$Customer3](https://docs.google.com/spreadsheets/d/1OEhzdMSlkGOaWyGKwdiAGlirsKKj9EN45Izn7kdKNTg/edit#gid=0&range=A1008) is in an active expansion deal and specifically asked about getting insights to (finally) work over >50 repos on 8/25. The AE (correctly, at the time) noted that would be out in 3.31. This is an enterprise-sized opportunity and they've been waiting for this limit to lift for 4+ months. 

* [$Prospect1](https://docs.google.com/spreadsheets/d/1OEhzdMSlkGOaWyGKwdiAGlirsKKj9EN45Izn7kdKNTg/edit#gid=0&range=A75) has [continually asked when we'll remove the repo limits](https://sourcegraph.slack.com/archives/C024QE1K0FP/p1623742523024800?thread_ts=1623703314.013800&cid=C024QE1K0FP) and [saw our Changelog + upgraded to 3.31](https://sourcegraph.slack.com/archives/C024QE1K0FP/p1630338287005800), and are thus expecting the ability to run insights over all their repositories now. They previously [ran into bugs in 3.30](https://sourcegraph.slack.com/archives/C024QE1K0FP/p1627337374069500) that we fixed in 3.31, but is also blocked until we merge this patch issue. TL;DR they're consistently trying to get insights over all repos (and being in the trial stage, can't "wait" as long as a regular customer). 

* [$Prospect2](https://docs.google.com/spreadsheets/d/1OEhzdMSlkGOaWyGKwdiAGlirsKKj9EN45Izn7kdKNTg/edit#gid=0&range=A11) has been asking about lifting the repo count limit [since May](https://sourcegraph.slack.com/archives/C0W2E592M/p1621961726047100). The deal is now in a late stage and it would be ideal to successfully deliver on this request. 

> Are the changes extremely minimal, well-tested, and low risk such that not testing as we do in a full release is OK?

From @coury-clark : 
_This change is very minimal and only fixes a panic caused at initialization. There are no behavior changes, and this unblocks one of the services from failing to start. There is more risk introduced in leaving this instability than applying without standard release testing_

> Is there some functionality completely broken that warrants redacting the prior release of Sourcegraph and advising users wait for the patch release?

Yes. Unfortunately, this indexing error prevents the worker service from starting up. The worker service powers the code insights backend. To borrow from @coury-clark again: 

_if at any point the queue reaches zero, the background job will immediately crash until it can be restarted into a state with a non-zero queue_

The queue will reach 0 every time your code insight(s) finishes running. It will crash, which then requires you to a) first, queue up more queries by creating more insights, and only then b) restarting the service (which will run the insights, then crash again). 

Given that most of our users won't have access or knowledge of how to restart the service on their instances, this essentially renders backend insights entirely broken. Worse, there's no error state – the insights will appear to be stuck "loading" forever, causing further customer frustration and negative product assumptions (around loading time). 

> This will interrupt our regular planned work and release cycle, taking one full working day of our time, and will take up all of our site admin's valuable time by asking them to upgrade or producing noise for them if they don't need to upgrade.
>
> Do you believe the changes are important enough to warrant this?

Yes. See the customer/prospects impacted above.

> Patch releases are a signal we can do something better to improve the quality of Sourcegraph. Have you already scheduled a call (or created a google doc) to perform a [retrospective](https://about.sourcegraph.com/retrospectives) and identify ways we can improve?

Yes. We will be completing an async retro on this item specifically over the next week [in this google doc](https://docs.google.com/document/d/1Dt389QMOMlxn5nEUSwiuLVSU_SN_xJ5rEc1yD0cf_jM/edit#heading=h.mfxpjh9k5amd). 

**For the [release captain](https://about.sourcegraph.com/handbook/engineering/releases#release-captain)** - after reviewing and approving this request:

- If there is [already an upcoming patch release](https://github.com/sourcegraph/sourcegraph/issues?q=is%3Aissue+label%3Arelease-tracking+), add the listed commits alongside a link to this issue
- If there is no upcoming patch release, create a new one:
  - Update [`dev/release/release-config.jsonc`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/dev/release/release-config.jsonc) with the patch release in `upcomingRelease` and `releaseDate` (and open a PR to `main` to update it)
  - `yarn release tracking:issues`
  - Add the listed commits alongside a link to this issue to the generated [release tracking issue](https://github.com/sourcegraph/sourcegraph/issues?q=is%3Aissue+label%3Arelease-tracking+)

Comment and close this issue once the relevant commit(s) have been cherry-picked into the release branch.

**Comment by Joelkw** [comment]
Reading this more closely today: 
> Is there some functionality completely broken that warrants redacting the prior release of Sourcegraph and advising users wait for the patch release?

I believe this is clear from my description above, but just to be explicit / not sure how we interpret "redacting" – this only affects customers (especially those listed) actively using code insights. There's no "redaction" needed in the sense of "we need to make sure _every instance_ on 3.31 gets onto 3.31.1" – [CEs know](https://sourcegraph.slack.com/archives/CU93UDUBV/p1630462059081300?thread_ts=1630446272.072800&cid=CU93UDUBV) which customers who are already on 3.31 who need to upgrade to 3.31.1 are. Said differently, "advising users wait for the patch release" is very important for any insights customers, but "redaction the prior release" is not important for customers who have already upgraded. cc @ggilmore

**Comment by felixfbecker** [comment]
Closing according to the instructions since the release is out, thank you @ggilmore for the quick turnaround!!

> Comment and close this issue once the relevant commit(s) have been cherry-picked into the release branch.
